core/src/test/java/de/unistuttgart/iaas/amyassist/amy/core/audio/sound/BasicSoundPlayerTest.java

```diff
@@ -0,0 +1,326 @@
+			}
+
+			Assertions.assertTrue(this.stopHookExecuted, "Did not run stop hook.");
+			Assertions.assertTrue(this.wasShouldStopWhenHookExecuted, "Should not stop yet.");
+			Assertions.assertEquals(StopReason.PLAYER_STOPPED, this.reason, "Wrong stop reason.");
+		}
+
+	}
+
+	/**
+	 * Test method for
+	 * {@link de.unistuttgart.iaas.amyassist.amy.core.audio.sound.BasicSoundPlayer#setOnStopHook(java.util.function.Consumer)}.
+	 * for {@link StopReason#OTHER}.
+	 * 
+	 * @throws InterruptedException
+	 *             When the thread gets interrupted
+	 * @throws IOException
+	 *             When an io error occurs
+	 */
+	@Test
+	void testSetOnStopPlayerOther() throws InterruptedException, IOException {
+		byte[] data = new byte[10020];
+
+		fillBuffer(data);
+
+		BasicSoundPlayer player = new BasicSoundPlayer(data, getFormat(), 1);
+
+		player.setOnStopHook(r -> {
+			this.stopHookExecuted = true;
+			this.wasShouldStopWhenHookExecuted = this.shouldStop;
+			this.reason = r;
+		});
+
+		try (AudioInputStream ais = player.getAudioStream()) {
+			player.start();
+
+			ais.read(data, 0, 10);
+			this.shouldStop = true;
+			Thread[] threads = new Thread[Thread.activeCount()];
+			Thread.enumerate(threads);
+			for (Thread t : threads) {
+				if (t.getName().equals("SoundPlayerThread")) {
+					t.interrupt();
+				}
+			}
+
+			while (player.isRunning()) {
+				// this is legitimate, deterministic use of sleep
+				Thread.sleep(100); // NOSONAR
+			}
+
+			Assertions.assertTrue(this.stopHookExecuted, "Did not run stop hook.");
+			Assertions.assertTrue(this.wasShouldStopWhenHookExecuted, "Should not stop yet.");
+			Assertions.assertEquals(StopReason.OTHER, this.reason, "Wrong stop reason.");
+		}
+
+	}
+
+	/**
+	 * Test method for
+	 * {@link de.unistuttgart.iaas.amyassist.amy.core.audio.sound.BasicSoundPlayer#setOnStopHook(java.util.function.Consumer)}.
+	 * when the player is already started.
+	 * 
+	 * @throws InterruptedException
+	 *             When the thread gets interrupted
+	 * @throws IOException
+	 *             When an io error occurs
+	 */
+	@Test
+	void testSetOnStopPlayerStarted() throws InterruptedException, IOException {
+		byte[] data = new byte[10020];
+
+		fillBuffer(data);
+
+		BasicSoundPlayer player = new BasicSoundPlayer(data, getFormat(), 1);
+
+		player.start();
+
+		Assertions.assertThrows(IllegalStateException.class,
+				() -> player.setOnStopHook(r -> this.stopHookExecuted = true));
+
+		try (AudioInputStream ais = player.getAudioStream()) {
+
+			readAll(data, ais);
+
+			while (player.isRunning()) {
+				// this is legitimate, deterministic use of sleep
+				Thread.sleep(100); // NOSONAR
+			}
+
+			Assertions.assertFalse(this.stopHookExecuted, "Stop hook should not be run.");
+		}
+
+	}
+
+	private void fillBuffer(byte[] buffer) {
+		for (int i = 0; i < buffer.length; i++) {
+			buffer[i] = (byte) i;
+		}
+	}
+
+	private int readAll(byte[] buffer, AudioInputStream stream) throws IOException {
```

Comment:
You could've used an implementation of the same concept from Apache Commons, like `IOUtils.toByteArray(is);`. Less potential for error. But now that you've already implemented it yourself, you can keep it 🙃

Reply:
Well. Didn't think about that.